File: #207_Course_Schedule/Solution.java
import java.util.*;

public class Solution {

	boolean[] isVisited;
	List<Integer>[] adj;

    public boolean canFinish(int numCourses, int[][] prerequisites) {
    	if (numCourses == 0 || prerequisites == null || prerequisites.length == 0) return true;
    	adj = (List<Integer>[]) new LinkedList[numCourses];
    	for (int i = 0; i < numCourses; i++)
    		adj[i] = (List<Integer>) new LinkedList();
    	isVisited = new boolean[numCourses];
    	for (int i = 0; i < prerequisites.length; i++) {
    		adj[prerequisites[i][1]].add(prerequisites[i][0]);
    	}
    	for (int i = 0; i < numCourses; i++) {
    		if (adj[i].isEmpty()) continue;
    		if (dfs(i)) return false;
    	}
    	return true;
    }

    // true if is visited before
    private boolean dfs(int i) {
    	if (isVisited[i]) return true;
    	isVisited[i] = true;
    	for (int neighbor : adj[i]) {
    		if (dfs(neighbor)) return true;
    	}
    	isVisited[i] = false;
    	return false;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[][] p = {
    		{0, 1}
    	};
    	System.out.println(sol.canFinish(2, p));
    }
}

File: #207_Course_Schedule/Solution.py
'''
Given a directed graph, find if the graph contains circle
For directed graph, don't need to explicitly handle the case of self loop and parallel edge
'''

class Solution(object):

	def __init__(self):
		self.isVisited = None
		self.onPath = None
		self.adj = None
		self.hasCycle = False

	def canFinish(self, numCourses, prerequisites):
		if not numCourses:
			return True
		self.isVisited = [False for x in range(numCourses)]
		self.onPath = [False for x in range(numCourses)]
		self.adj = [[] for x in range(numCourses)]
		for pre in prerequisites:
			self.adj[pre[1]].append(pre[0])
		for x in range(numCourses):
			if not self.hasCycle and not self.isVisited[x]:
				self.dfs(x)
		return not self.hasCycle

	def dfs(self, v):
		self.isVisited[v] = True
		self.onPath[v] = True
		for w in self.adj[v]:
			if not self.isVisited[w]:
				self.dfs(w)
			elif self.onPath[w]:
				self.hasCycle = True
				return
		self.onPath[v] = False

sol = Solution()
print sol.canFinish(2, [[1, 0], [0, 1]])